test/micro/org/openjdk/bench/vm/gc/system_gc/AllDead.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2017, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.vm.gc.system_gc;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+
+import java.util.ArrayList;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.SingleShotTime)
```

Comment:
I had a discussion with Stefan about the usefulness of using single-shot mode (i.e. no warmup, no iterations).
This can make algorithms (like Serial GC MSC) look better than they are as parts of the algorithm may be dependent on existing dead objects (i.e. it needs to iterate over dead objects to find marked objects). However setting up a "fair" dead object distribution isn't that easy. One could simply just assume the previous' GC's dead objects are representative "fair", but one can argue endlessly about it.

Just something to consider when interpreting the results.

Reply:
I agree, generating good and fair pre-conditions for these tests is not straight forward. As mentioned offline I find going with a new JVM for each run to be most fair, to generate scenarios where there are dead objects in the old generation we could have tests where the setup does a `System.gc()` to move parts of the "heap under test" to old. This could be interesting, but could potentially also lead to more variance.